// @glimmer/syntax/dist/commonjs/es2017/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.printLiteral = exports.isLiteral = exports.AST = exports.SyntaxError = exports.print = exports.Walker = exports.traverse = exports.cannotReplaceOrRemoveInKeyHandlerYet = exports.cannotReplaceNode = exports.cannotRemoveNode = exports.TraversalError = exports.builders = exports.preprocess = undefined;

var _tokenizerEventHandlers = require('./lib/parser/tokenizer-event-handlers');

Object.defineProperty(exports, 'preprocess', {
  enumerable: true,
  get: function () {
    return _tokenizerEventHandlers.preprocess;
  }
});

var _builders = require('./lib/builders');

Object.defineProperty(exports, 'builders', {
  enumerable: true,
  get: function () {
    return _interopRequireDefault(_builders).default;
  }
});

var _errors = require('./lib/traversal/errors');

Object.defineProperty(exports, 'TraversalError', {
  enumerable: true,
  get: function () {
    return _interopRequireDefault(_errors).default;
  }
});
Object.defineProperty(exports, 'cannotRemoveNode', {
  enumerable: true,
  get: function () {
    return _errors.cannotRemoveNode;
  }
});
Object.defineProperty(exports, 'cannotReplaceNode', {
  enumerable: true,
  get: function () {
    return _errors.cannotReplaceNode;
  }
});
Object.defineProperty(exports, 'cannotReplaceOrRemoveInKeyHandlerYet', {
  enumerable: true,
  get: function () {
    return _errors.cannotReplaceOrRemoveInKeyHandlerYet;
  }
});

var _traverse = require('./lib/traversal/traverse');

Object.defineProperty(exports, 'traverse', {
  enumerable: true,
  get: function () {
    return _interopRequireDefault(_traverse).default;
  }
});

var _walker = require('./lib/traversal/walker');

Object.defineProperty(exports, 'Walker', {
  enumerable: true,
  get: function () {
    return _interopRequireDefault(_walker).default;
  }
});

var _print = require('./lib/generation/print');

Object.defineProperty(exports, 'print', {
  enumerable: true,
  get: function () {
    return _interopRequireDefault(_print).default;
  }
});

var _syntaxError = require('./lib/errors/syntax-error');

Object.defineProperty(exports, 'SyntaxError', {
  enumerable: true,
  get: function () {
    return _interopRequireDefault(_syntaxError).default;
  }
});

var _utils = require('./lib/utils');

Object.defineProperty(exports, 'isLiteral', {
  enumerable: true,
  get: function () {
    return _utils.isLiteral;
  }
});
Object.defineProperty(exports, 'printLiteral', {
  enumerable: true,
  get: function () {
    return _utils.printLiteral;
  }
});

var _nodes = require('./lib/types/nodes');

var AST = _interopRequireWildcard(_nodes);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
// AST


exports.AST = AST;
