// eslint-plugin-ember/lib/index.js
'use strict';

/* eslint-disable global-require */
module.exports = {
  rules: {
    'alias-model-in-controller': require('./rules/alias-model-in-controller'),
    'avoid-leaking-state-in-components': require('./rules/avoid-leaking-state-in-components'),
    'avoid-leaking-state-in-ember-objects': require('./rules/avoid-leaking-state-in-ember-objects'),
    'avoid-using-needs-in-controllers': require('./rules/avoid-using-needs-in-controllers'),
    'classic-decorator-hooks': require('./rules/classic-decorator-hooks'),
    'classic-decorator-no-classic-methods': require('./rules/classic-decorator-no-classic-methods'),
    'closure-actions': require('./rules/closure-actions'),
    'computed-property-getters': require('./rules/computed-property-getters'),
    'jquery-ember-run': require('./rules/jquery-ember-run'),
    'local-modules': require('./rules/local-modules'),
    'named-functions-in-promises': require('./rules/named-functions-in-promises'),
    'new-module-imports': require('./rules/new-module-imports'),
    'no-arrow-function-computed-properties': require('./rules/no-arrow-function-computed-properties'),
    'no-attrs-in-components': require('./rules/no-attrs-in-components'),
    'no-attrs-snapshot': require('./rules/no-attrs-snapshot'),
    'no-capital-letters-in-routes': require('./rules/no-capital-letters-in-routes'),
    'no-deeply-nested-dependent-keys-with-each': require('./rules/no-deeply-nested-dependent-keys-with-each'),
    'no-duplicate-dependent-keys': require('./rules/no-duplicate-dependent-keys'),
    'no-ember-super-in-es-classes': require('./rules/no-ember-super-in-es-classes'),
    'no-ember-testing-in-module-scope': require('./rules/no-ember-testing-in-module-scope'),
    'no-empty-attrs': require('./rules/no-empty-attrs'),
    'no-function-prototype-extensions': require('./rules/no-function-prototype-extensions'),
    'no-get-properties': require('./rules/no-get-properties'),
    'no-get': require('./rules/no-get'),
    'no-global-jquery': require('./rules/no-global-jquery'),
    'no-incorrect-calls-with-inline-anonymous-functions': require('./rules/no-incorrect-calls-with-inline-anonymous-functions'),
    'no-invalid-debug-function-arguments': require('./rules/no-invalid-debug-function-arguments'),
    'no-jquery': require('./rules/no-jquery'),
    'no-new-mixins': require('./rules/no-new-mixins'),
    'no-observers': require('./rules/no-observers'),
    'no-old-shims': require('./rules/no-old-shims'),
    'no-on-calls-in-components': require('./rules/no-on-calls-in-components'),
    'no-proxies': require('./rules/no-proxies'),
    'no-restricted-resolver-tests': require('./rules/no-restricted-resolver-tests'),
    'no-side-effects': require('./rules/no-side-effects'),
    'no-test-and-then': require('./rules/no-test-and-then'),
    'no-test-import-export': require('./rules/no-test-import-export'),
    'no-unnecessary-index-route': require('./rules/no-unnecessary-index-route'),
    'no-unnecessary-route-path-option': require('./rules/no-unnecessary-route-path-option'),
    'no-unnecessary-service-injection-argument': require('./rules/no-unnecessary-service-injection-argument'),
    'no-volatile-computed-properties': require('./rules/no-volatile-computed-properties'),
    'order-in-components': require('./rules/order-in-components'),
    'order-in-controllers': require('./rules/order-in-controllers'),
    'order-in-models': require('./rules/order-in-models'),
    'order-in-routes': require('./rules/order-in-routes'),
    'require-computed-macros': require('./rules/require-computed-macros'),
    'require-computed-property-dependencies': require('./rules/require-computed-property-dependencies'),
    'require-return-from-computed': require('./rules/require-return-from-computed'),
    'require-super-in-init': require('./rules/require-super-in-init'),
    'route-path-style': require('./rules/route-path-style'),
    'routes-segments-snake-case': require('./rules/routes-segments-snake-case'),
    'use-brace-expansion': require('./rules/use-brace-expansion'),
    'use-ember-data-rfc-395-imports': require('./rules/use-ember-data-rfc-395-imports'),
    'use-ember-get-and-set': require('./rules/use-ember-get-and-set'),
  },
  configs: {
    base: require('./config/base'),
    recommended: require('./config/recommended'),
  },
  utils: {
    ember: require('./utils/ember'),
    utils: require('./utils/utils'),
  },
};
