// @glimmer/di/dist/commonjs/es2017/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _container = require('./container');

Object.defineProperty(exports, 'Container', {
  enumerable: true,
  get: function () {
    return _interopRequireDefault(_container).default;
  }
});

var _registry = require('./registry');

Object.defineProperty(exports, 'Registry', {
  enumerable: true,
  get: function () {
    return _interopRequireDefault(_registry).default;
  }
});

var _owner = require('./owner');

Object.defineProperty(exports, 'getOwner', {
  enumerable: true,
  get: function () {
    return _owner.getOwner;
  }
});
Object.defineProperty(exports, 'setOwner', {
  enumerable: true,
  get: function () {
    return _owner.setOwner;
  }
});
Object.defineProperty(exports, 'OWNER', {
  enumerable: true,
  get: function () {
    return _owner.OWNER;
  }
});

var _specifier = require('./specifier');

Object.defineProperty(exports, 'isSpecifierStringAbsolute', {
  enumerable: true,
  get: function () {
    return _specifier.isSpecifierStringAbsolute;
  }
});
Object.defineProperty(exports, 'isSpecifierObjectAbsolute', {
  enumerable: true,
  get: function () {
    return _specifier.isSpecifierObjectAbsolute;
  }
});
Object.defineProperty(exports, 'serializeSpecifier', {
  enumerable: true,
  get: function () {
    return _specifier.serializeSpecifier;
  }
});
Object.defineProperty(exports, 'deserializeSpecifier', {
  enumerable: true,
  get: function () {
    return _specifier.deserializeSpecifier;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
